File: js/Vec2.js
// Vec2 — 2D vector utility
class Vec2 {
  constructor(x = 0, y = 0) { this.x = x; this.y = y; }
  clone()            { return new Vec2(this.x, this.y); }
  set(x, y)         { this.x = x; this.y = y; return this; }
  add(v)            { return new Vec2(this.x + v.x, this.y + v.y); }
  sub(v)            { return new Vec2(this.x - v.x, this.y - v.y); }
  scale(s)          { return new Vec2(this.x * s, this.y * s); }
  length()          { return Math.sqrt(this.x * this.x + this.y * this.y); }
  normalized()      { const l = this.length(); return l > 0 ? this.scale(1 / l) : new Vec2(); }
  dot(v)            { return this.x * v.x + this.y * v.y; }
  addSelf(v)        { this.x += v.x; this.y += v.y; return this; }
  scaleSelf(s)      { this.x *= s; this.y *= s; return this; }
}

File: references/dyad_motorcontrol/app/src/main/java/com/nirsense/webswinger/engine/Vec2.kt
package com.nirsense.webswinger.engine

/**
 * Simple 2D vector class used throughout the game engine.
 */
data class Vec2(var x: Float = 0f, var y: Float = 0f) {
    operator fun plus(other: Vec2) = Vec2(x + other.x, y + other.y)
    operator fun minus(other: Vec2) = Vec2(x - other.x, y - other.y)
    operator fun times(s: Float) = Vec2(x * s, y * s)

    fun length(): Float = Math.sqrt((x * x + y * y).toDouble()).toFloat()

    fun normalized(): Vec2 {
        val len = length()
        return if (len > 0.0001f) Vec2(x / len, y / len) else Vec2(0f, 0f)
    }

    fun dot(other: Vec2): Float = x * other.x + y * other.y

    fun set(nx: Float, ny: Float) { x = nx; y = ny }
    fun set(other: Vec2) { x = other.x; y = other.y }
}
